ocs_ci/ocs/provider_client_upgrade.py

```diff
@@ -0,0 +1,227 @@
+                        resource_name=p.get()["metadata"]["name"], column="AGE"
+                    )
+                    self.post_upgrade_data["pod_status"] = pod_obj.get_resource_status(
+                        resource_name=p.get()["metadata"]["name"]
+                    )
+
+        # get pre-upgrade csv
+        csv_objs = CSV(namespace=self.namespace)
+        for csv in csv_objs.get()["items"]:
+            if self.operator_name in csv["metadata"]["name"]:
+                csv_obj = CSV(
+                    namespace=self.namespace, resource_name=csv["metadata"]["name"]
+                )
+                if self.upgrade_phase == "pre_upgrade":
+                    self.pre_upgrade_data["version"] = csv_obj.get_resource(
+                        resource_name=csv_obj.resource_name, column="VERSION"
+                    )
+                    try:
+                        self.pre_upgrade_data["version"]
+                    except KeyError:
+                        log.error(
+                            f"Couldn't capture Pre-upgrade CSV version for {self.operator_name}"
+                        )
+                if self.upgrade_phase == "post_upgrade":
+                    if self.upgrade_version in csv["metadata"]["name"]:
+                        self.post_upgrade_data["version"] = csv_obj.get_resource(
+                            resource_name=csv_obj.resource_name, column="VERSION"
+                        )
+                    try:
+                        self.post_upgrade_data["version"]
+                    except KeyError:
+                        log.error(
+                            f"Couldn't capture Post upgrade CSV version for {self.operator_name}"
+                        )
+        # Make sure all csvs are in succeeded state
+        check_all_csvs_are_succeeded(namespace=self.namespace)
+
+
+class OperatorUpgrade(ProviderUpgrade):
+    """
+    A class to handle installed operators on provider upgrades
+
+    """
+
+    def __init__(self):
+        self.storage_clients = StorageClient()
+        self.metallb_installer_obj = MetalLBInstaller()
+        self.cnv_installer_obj = CNVInstaller()
+        self.acm_hub_upgrade_obj = ACMUpgrade()
+
+    def collect_pre_upgrade_data(self):
+        # Collect some pre-upgrade data for comparision after the upgrade
+        self.metallb_pod_name_pattern = "metallb-operator"
+        self.collect_data(self.metallb_pod_name_pattern)
+        assert (
+            self.pre_upgrade_data.get("pod_status", "") == "Running"
+        ), "metallb operator pod is not in Running status"
+        # Collect some pre-upgrade data for comparision after the upgrade
+        self.cnv_pod_name_pattern = "klusterlet"
+        self.collect_data(self.cnv_pod_name_pattern)
+        assert (
+            self.pre_upgrade_data.get("pod_status", "") == "Running"
+        ), "klusterlet pods are not in Running status"
+        self.run_upgrade()
+        ocs_upgrade.run_ocs_upgrade()
```

Comment:
not clear why we do run_upgrade and then ocs_upgrade.run_ocs_upgrade. ProviderUpgrade should have include OCS upgrade as well. The design can be improved including ocs provider update to ProviderUpgrade

Reply:
Will update this in next pr